instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java

```diff
@@ -0,0 +1,362 @@
+            .and(not(isAbstract()))
+            .and(takesArguments(4))
+            .and(takesArgument(0, named("org.apache.hc.core5.http.HttpHost")))
+            .and(takesArgument(1, named("org.apache.hc.core5.http.ClassicHttpRequest")))
+            .and(takesArgument(2, named("org.apache.hc.core5.http.protocol.HttpContext")))
+            .and(takesArgument(3, named("org.apache.hc.core5.http.io.HttpClientResponseHandler"))),
+        this.getClass().getName() + "$RequestWithHostAndContextAndHandlerAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class RequestAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void methodEnter(
+        @Advice.Argument(0) ClassicHttpRequest request,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      Context parentContext = currentContext();
+      if (!instrumenter().shouldStart(parentContext, request)) {
+        return;
+      }
+
+      context = instrumenter().start(parentContext, request);
+      scope = context.makeCurrent();
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void methodExit(
+        @Advice.Argument(0) ClassicHttpRequest request,
+        @Advice.Return Object result,
+        @Advice.Thrown Throwable throwable,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      if (scope == null) {
+        return;
+      }
+
+      scope.close();
+      ApacheHttpClientHelper.doMethodExit(context, request, result, throwable);
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class RequestWithHandlerAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void methodEnter(
+        @Advice.Argument(0) ClassicHttpRequest request,
+        @Advice.Argument(value = 1, readOnly = false) HttpClientResponseHandler<?> handler,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      Context parentContext = currentContext();
+      if (!instrumenter().shouldStart(parentContext, request)) {
+        return;
+      }
+
+      context = instrumenter().start(parentContext, request);
+      scope = context.makeCurrent();
+
+      // Wrap the handler so we capture the status code
+      if (handler != null) {
+        handler =
+            new WrappingStatusSettingResponseHandler<>(context, parentContext, request, handler);
+      }
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void methodExit(
+        @Advice.Argument(0) ClassicHttpRequest request,
+        @Advice.Return Object result,
+        @Advice.Thrown Throwable throwable,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      if (scope == null) {
+        return;
+      }
+
+      scope.close();
```

Comment:
Can closing the scope fail? If so, should put `ApacheHttpClientHelper.doMethodExit` in a `finally`?

Reply:
No. If it finds the scope is not part of the current context, it logs a "fine" message saying close is ignored